### PR TITLE
[ network ] add binary support for socket + fix recvAll

### PR DESCRIPTION
**Pull request by octeep**
This add `recvAllBytes`, `recvBytes` and `sendBytes`, which are variants of `recvAll`, `recv`, and `send` but uses `List Bits8` instead of `String`.

This PR tackles #2211, #2307, and #2282

### CODE REVIEW

File: libs/network/Network/Socket.idr [hunk 1]
```diff
@@ -164,6 +166,12 @@ recv sock len = do
            freeRecvStruct (RSPtr recv_struct_ptr)
            pure $ Right (payload, recv_res)
 
+recvAllRec : (Monoid a, HasIO io) => io (Either SocketError a) -> List a -> io (Either SocketError a)
+recvAllRec recv_from_socket acc = case !recv_from_socket of
+  Left 0 => pure (Right $ concat $ reverse acc)
+  Left c => pure (Left c)
+  Right str => recvAllRec recv_from_socket (str :: acc)
```
Comment (gallais) [inline libs/network/Network/Socket.idr:173]:
```suggestion
recvAllRec : (Monoid a, HasIO io) => io (Either SocketError a) -> SnocList a -> io (Either SocketError a)
recvAllRec recv_from_socket acc = case !recv_from_socket of
  Left 0 => pure (Right $ concat acc)
  Left c => pure (Left c)
  Right str => recvAllRec recv_from_socket (acc :< str)
```

File: libs/network/Network/Socket.idr [hunk 2]
```diff
@@ -172,16 +180,7 @@ recv sock len = do
...
-  where
-    covering
-    recvRec : Socket -> List String -> ByteLength -> io (Either SocketError String)
-    recvRec sock acc n = do res <- recv sock n
-                            case res of
-                              Left c => pure (Left c)
-                              Right (str, res) => let n' = min (n * 2) 65536 in
-                                                  if res < n then pure (Right $ concat $ reverse $ str :: acc)
-                                                  else recvRec sock (str :: acc) n'
+recvAll sock = recvAllRec {a=String} (mapSnd fst <$> recv sock 65536) []
```
Comment (gallais) [inline libs/network/Network/Socket.idr:183]:
```suggestion
recvAll sock = recvAllRec {a=String} (mapSnd fst <$> recv sock 65536) [<]
```

File: libs/network/Network/Socket.idr [hunk 3]
```diff
@@ -243,3 +242,55 @@ recvFrom sock bl = do
...
+
+||| Receive all the remaining data on the specified socket.
+|||
+||| Returns on failure a `SocketError`
+||| Returns on success the payload `List Bits8`
+|||
+||| @sock The socket on which to receive the message.
+export
+recvAllBytes : HasIO io => (sock : Socket) -> io (Either SocketError (List Bits8))
+recvAllBytes sock = recvAllRec {a=List Bits8} (recvBytes sock 65536) []
```
Comment (gallais) [inline libs/network/Network/Socket.idr:296]:
```suggestion
recvAllBytes sock = recvAllRec {a=List Bits8} (recvBytes sock 65536) [<]
```

**Review by gallais — COMMENTED** [review]
Style-wise, I'd use a snoclist instead of a list+reverse on base case.

**Comment by octeep** [comment]
> Style-wise, I'd use a snoclist instead of a list+reverse on base case.

Just merged your suggestions. Is there anything else you'd like to add?

**Comment by gallais** [comment]
Nope, I was just waiting for a successful build. :ship:

**Pull request merged**